### PR TITLE
[mlir][spirv][cf] legalize block arguments when convert cf to spirv

### LINKED ISSUE #70813: [mlir][spirv] type mismatch for bb argument when convert-cf-to-spirv
**Command line**
  mlir-opt --convert-cf-to-spirv 

**Input mlir**
```
module {
  func.func @main_graph(%arg0: memref<3xf32>, %arg1: memref<3xf32>) -> memref<3xf32> {
    %alloc = memref.alloc() {alignment = 16 : i64} : memref<3xf32>
    %c0 = arith.constant 0 : index
    %c3 = arith.constant 3 : index
    %c1 = arith.constant 1 : index
    cf.br ^bb1(%c0 : index)
  ^bb1(%0: index):  // 2 preds: ^bb0, ^bb2
    %1 = arith.cmpi slt, %0, %c3 : index
    cf.cond_br %1, ^bb2, ^bb3
  ^bb2:  // pred: ^bb1
    %2 = memref.load %arg0[%0] : memref<3xf32>
    %3 = memref.load %arg1[%0] : memref<3xf32>
    %4 = arith.addf %2, %3 : f32
    memref.store %4, %alloc[%0] : memref<3xf32>
    %5 = arith.addi %0, %c1 : index
    cf.br ^bb1(%5 : index)
  ^bb3:  // pred: ^bb1
    return %alloc : memref<3xf32>
  }
}
```

**Output**
 error: type mismatch for bb argument #0 of successor #0
    cf.br ^bb1(%c0 : index)

### PR DESCRIPTION
**Pull request by python3kgae**
When converting branches, legalize target block arguments first.

Fixes llvm#70813

### CODE REVIEW

File: mlir/lib/Conversion/ControlFlowToSPIRV/ControlFlowToSPIRV.cpp [hunk 1]
```diff
@@ -28,20 +28,23 @@
...
+                                   const TypeConverter &converter) {
+  auto builder = OpBuilder::atBlockBegin(&block);
+  for (unsigned i = 0; i < block.getNumArguments(); ++i) {
+    const auto arg = block.getArgument(i);
+    if (converter.isLegal(arg.getType()))
+      continue;
+    unsigned argNum = arg.getArgNumber();
+    Location loc = arg.getLoc();
+    Type ty = arg.getType();
+    Type newTy = converter.convertType(ty);
```
Comment (kuhar) [inline mlir/lib/Conversion/ControlFlowToSPIRV/ControlFlowToSPIRV.cpp:42]: I think we need to handle potential conversion failures here.

File: mlir/lib/Conversion/ControlFlowToSPIRV/ControlFlowToSPIRV.cpp [hunk 2]
```diff
@@ -28,20 +28,23 @@
...
+    Value newArg = block.insertArgument(argNum, newTy, loc);
+    auto cast = builder.create<UnrealizedConversionCastOp>(loc, ty, newArg);
```
Comment (kuhar) [inline mlir/lib/Conversion/ControlFlowToSPIRV/ControlFlowToSPIRV.cpp:44]: Shouldn't we use something that relies on the 'source materialization' registered in the type converter? (`materializeSourceConversion`?)

File: mlir/lib/Conversion/ControlFlowToSPIRV/ControlFlowToSPIRV.cpp (hunk 1 above)
Comment (python3kgae) [reply mlir/lib/Conversion/ControlFlowToSPIRV/ControlFlowToSPIRV.cpp:41]: Fixed.

File: mlir/lib/Conversion/ControlFlowToSPIRV/ControlFlowToSPIRV.cpp (hunk 2 above)
Comment (python3kgae) [reply mlir/lib/Conversion/ControlFlowToSPIRV/ControlFlowToSPIRV.cpp]: Switched to materializeSourceConversion.

**Review by kuhar — APPROVED** [review]
LGTM, thanks for taking care of this!

**Pull request merged**